agent/unit/graph.go

```diff
@@ -0,0 +1,81 @@
+package unit
+
+import "sync"
+
+// Graph is an bidirectional adjacency list representation of a graph.
+// It is considered bidirectional instead of undirected, because we distinguish
+// between forward and reverse edges. Wanting and being wanted by other units
+// are related but different concepts that have different graph traversal implications
+// when Units update their status. Adding one of these directions necessarily adds
+// the other to the complementary unit.
+//
+// Graph vertices often have their own attributes specific to the problem domain.
+// In this case we need to distinguish between different edge types to represent
+// the different relationships between units.
+type Graph[EdgeType, VertexType comparable] struct {
+	mu                   sync.RWMutex
+	adjacencyList        map[VertexType]map[VertexType]EdgeType
+	reverseAdjacencyList map[VertexType]map[VertexType]EdgeType
+}
+
+// Edge is a convenience type for representing an edge in the graph.
+// It encapsulates the from and to vertices and the edge type itself.
+type Edge[EdgeType, VertexType comparable] struct {
+	From VertexType
+	To   VertexType
+	Edge EdgeType
+}
+
+func NewGraph[EdgeType, VertexType comparable]() *Graph[EdgeType, VertexType] {
+	return &Graph[EdgeType, VertexType]{
+		adjacencyList:        make(map[VertexType]map[VertexType]EdgeType),
+		reverseAdjacencyList: make(map[VertexType]map[VertexType]EdgeType),
+	}
+}
+
+// AddEdge adds an edge to the graph. It initializes the adjacency lists if they don't exist
+// and adds the edge to both the adjacency list and the reverse adjacency list.
+func (g *Graph[EdgeType, VertexType]) AddEdge(from, to VertexType, edge EdgeType) {
```

Comment:
Should we guard against cycles and self-reference? If not here, perhaps a level higher?